File: lib/vault.ts
import fs from 'fs';
import os from 'os';
import path from 'path';

const APP_CONFIG_DIR = 'com.leaflyte.desktop';

function desktopConfigDir(): string {
  const home = os.homedir();
  if (process.platform === 'darwin') {
    return path.join(home, 'Library', 'Application Support', APP_CONFIG_DIR);
  }
  if (process.platform === 'win32') {
    return path.join(
      process.env.APPDATA || path.join(home, 'AppData', 'Roaming'),
      APP_CONFIG_DIR
    );
  }
  return path.join(home, '.local', 'share', APP_CONFIG_DIR);
}

/** Same persisted path the Tauri desktop app uses (Application Support/com.leaflyte.desktop/vault-path.txt). */
function loadPersistedVaultPath(): string | null {
  try {
    const raw = fs.readFileSync(path.join(desktopConfigDir(), 'vault-path.txt'), 'utf8').trim();
    if (!raw) return null;
    const resolved = path.resolve(raw);
    if (fs.existsSync(resolved) && fs.statSync(resolved).isDirectory()) {
      return resolved;
    }
  } catch {
    // no persisted vault yet
  }
  return null;
}

/** Active vault root for Next.js API routes (capture, dev file APIs). */
export function getVaultDir(): string {
  if (process.env.VAULT_DIR) {
    return path.resolve(process.env.VAULT_DIR);
  }
  const persisted = loadPersistedVaultPath();
  if (persisted) return persisted;
  return path.join(process.cwd(), 'vault');
}

/** @deprecated Prefer getVaultDir() — capture and file APIs resolve dynamically. */
export const VAULT_DIR = getVaultDir();

const IGNORE = new Set(['.git', 'node_modules', '.DS_Store', '.leaflyte-index']);

function loadGitignore(): string[] {
  const vaultDir = getVaultDir();
  try {
    const raw = fs.readFileSync(path.join(vaultDir, '.gitignore'), 'utf8');
    return raw
      .split('\n')
      .map((l) => l.trim())
      .filter((l) => l && !l.startsWith('#'))
      .map((l) => l.replace(/^\//, ''));
  } catch {
    return [];
  }
}

function gitignoreMatch(rel: string, pattern: string): boolean {
  const normalized = rel.replace(/^\/+/, '');
  if (pattern.includes('*')) {
    if (pattern.startsWith('*')) return normalized.endsWith(pattern.slice(1));
    if (pattern.endsWith('*')) return normalized.startsWith(pattern.slice(0, -1));
  }
  if (normalized === pattern || normalized.endsWith(`/${pattern}`)) return true;
  return normalized.split('/').some((seg) => seg === pattern);
}

function isGitignored(rel: string, patterns: string[]): boolean {
  return patterns.some((p) => gitignoreMatch(rel, p));
}

export interface TreeNode {
  name: string;
  path: string; // relative to vault root, posix-style
  type: 'file' | 'folder';
  children?: TreeNode[];
}

/** Resolve a relative vault path safely, refusing to escape the vault root. */
export function resolveSafe(relPath: string): string {
  const vaultDir = getVaultDir();
  const cleaned = relPath.replace(/^\/+/, '');
  const abs = path.resolve(vaultDir, cleaned);
  if (abs !== vaultDir && !abs.startsWith(vaultDir + path.sep)) {
    throw new Error('Path escapes vault directory');
  }
  return abs;
}

export function ensureVault() {
  const vaultDir = getVaultDir();
  if (!fs.existsSync(vaultDir)) {
    fs.mkdirSync(vaultDir, { recursive: true });
  }
}

function toPosix(p: string) {
  return p.split(path.sep).join('/');
}

export function listTree(dir?: string, gitignore?: string[]): TreeNode[] {
  const vaultDir = getVaultDir();
  const root = dir ?? vaultDir;
  const ignore = gitignore ?? loadGitignore();
  ensureVault();
  const entries = fs.readdirSync(root, { withFileTypes: true });
  const nodes: TreeNode[] = [];

  for (const entry of entries) {
    if (IGNORE.has(entry.name) || entry.name.startsWith('.')) continue;
    const abs = path.join(root, entry.name);
    const rel = toPosix(path.relative(vaultDir, abs));
    if (isGitignored(rel, ignore)) continue;

    if (entry.isDirectory()) {
      nodes.push({
        name: entry.name,
        path: rel,
        type: 'folder',
        children: listTree(abs, ignore)
      });
    } else {
      nodes.push({ name: entry.name, path: rel, type: 'file' });
    }
  }

  nodes.sort((a, b) => {
    if (a.type !== b.type) return a.type === 'folder' ? -1 : 1;
    return a.name.localeCompare(b.name);
  });

  return nodes;
}

export function readFile(relPath: string): string {
  const abs = resolveSafe(relPath);
  return fs.readFileSync(abs, 'utf8');
}

export function writeFile(relPath: string, content: string) {
  const abs = resolveSafe(relPath);
  fs.mkdirSync(path.dirname(abs), { recursive: true });
  fs.writeFileSync(abs, content, 'utf8');
}

export function deleteFile(relPath: string) {
  const abs = resolveSafe(relPath);
  const stat = fs.statSync(abs);
  if (stat.isDirectory()) {
    fs.rmSync(abs, { recursive: true, force: true });
  } else {
    fs.rmSync(abs);
  }
}

export function exists(relPath: string): boolean {
  try {
    return fs.existsSync(resolveSafe(relPath));
  } catch {
    return false;
  }
}

function toPosixRel(relPath: string) {
  return relPath.replace(/^\/+/, '').split(path.sep).join('/');
}

/** Move a file or folder to a new vault-relative path. */
export function movePath(fromRel: string, toRel: string) {
  const from = toPosixRel(fromRel);
  const to = toPosixRel(toRel);
  if (!from || !to) throw new Error('from and to are required');
  if (from === to) return to;

  const fromAbs = resolveSafe(from);
  const toAbs = resolveSafe(to);

  if (!fs.existsSync(fromAbs)) throw new Error('Source not found');
  if (fs.existsSync(toAbs)) throw new Error('A file or folder already exists at the destination');
  if (toAbs === fromAbs || toAbs.startsWith(fromAbs + path.sep)) {
    throw new Error('Cannot move a folder into itself');
  }

  fs.mkdirSync(path.dirname(toAbs), { recursive: true });
  const isDir = fs.statSync(fromAbs).isDirectory();
  try {
    fs.renameSync(fromAbs, toAbs);
  } catch (err) {
    if (!isDir) throw err;
    fs.cpSync(fromAbs, toAbs, { recursive: true, errorOnExist: true });
    fs.rmSync(fromAbs, { recursive: true, force: true });
  }
  return to;
}

/** Flat list of every file (not folders) in the vault, relative posix paths. */
export function listAllFiles(dir?: string): string[] {
  const vaultDir = getVaultDir();
  const root = dir ?? vaultDir;
  const gitignore = loadGitignore();
  const out: string[] = [];
  const walk = (d: string) => {
    for (const entry of fs.readdirSync(d, { withFileTypes: true })) {
      if (IGNORE.has(entry.name) || entry.name.startsWith('.')) continue;
      const abs = path.join(d, entry.name);
      const rel = toPosix(path.relative(vaultDir, abs));
      if (isGitignored(rel, gitignore)) continue;
      if (entry.isDirectory()) {
        walk(abs);
      } else {
        out.push(rel);
      }
    }
  };
  ensureVault();
  walk(root);
  return out;
}

File: src-tauri/src/vault.rs
use serde::Serialize;
use std::fs;
use std::path::{Component, Path, PathBuf};
use std::sync::{Arc, Mutex};

const IGNORE: &[&str] = &[".git", "node_modules", ".DS_Store", ".leaflyte-index"];
const APP_CONFIG_DIR: &str = "com.leaflyte.desktop";

#[derive(Serialize, Clone)]
pub struct TreeNode {
    pub name: String,
    pub path: String,
    #[serde(rename = "type")]
    pub node_type: String,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub children: Option<Vec<TreeNode>>,
}

#[derive(Serialize)]
pub struct WriteResult {
    pub created: bool,
}

pub struct VaultState {
    pub root: Mutex<PathBuf>,
}

impl VaultState {
    pub fn new() -> Self {
        Self {
            root: Mutex::new(initial_root()),
        }
    }
}

pub fn app_config_dir() -> Option<PathBuf> {
    dirs::data_dir().map(|dir| dir.join(APP_CONFIG_DIR))
}

fn config_dir() -> Option<PathBuf> {
    app_config_dir()
}

fn persisted_vault_config_path() -> Option<PathBuf> {
    config_dir().map(|dir| dir.join("vault-path.txt"))
}

fn load_persisted_vault_path() -> Option<PathBuf> {
    let file = persisted_vault_config_path()?;
    let raw = fs::read_to_string(file).ok()?;
    let trimmed = raw.trim();
    if trimmed.is_empty() {
        return None;
    }
    let path = PathBuf::from(trimmed);
    if path.is_dir() { Some(path) } else { None }
}

fn save_persisted_vault_path(path: &Path) -> Result<(), String> {
    let dir = config_dir().ok_or("Could not resolve app config directory")?;
    fs::create_dir_all(&dir).map_err(|e| e.to_string())?;
    fs::write(
        dir.join("vault-path.txt"),
        path.to_string_lossy().as_bytes(),
    )
    .map_err(|e| e.to_string())
}

#[cfg(unix)]
fn same_path(a: &Path, b: &Path) -> bool {
    use std::os::unix::fs::MetadataExt;
    match (fs::metadata(a), fs::metadata(b)) {
        (Ok(left), Ok(right)) => left.dev() == right.dev() && left.ino() == right.ino(),
        _ => false,
    }
}

#[cfg(not(unix))]
fn same_path(a: &Path, b: &Path) -> bool {
    a == b
}

fn initial_root() -> PathBuf {
    if cfg!(debug_assertions) {
        return default_dev_root();
    }
    if let Some(persisted) = load_persisted_vault_path() {
        return persisted;
    }
    default_documents_root()
}

fn default_dev_root() -> PathBuf {
    let p = PathBuf::from(env!("CARGO_MANIFEST_DIR")).join("../vault");
    let _ = fs::create_dir_all(&p);
    fs::canonicalize(&p).unwrap_or(p)
}

fn default_documents_root() -> PathBuf {
    let base = dirs::document_dir()
        .or_else(dirs::home_dir)
        .unwrap_or_else(|| PathBuf::from("."));
    let p = base.join("Leaflyte");
    let _ = fs::create_dir_all(&p);
    fs::canonicalize(&p).unwrap_or(p)
}

pub fn root_of(state: &VaultState) -> Result<PathBuf, String> {
    Ok(state.root.lock().map_err(|e| e.to_string())?.clone())
}

fn set_root(state: &VaultState, path: PathBuf) -> Result<(), String> {
    *state.root.lock().map_err(|e| e.to_string())? = path;
    Ok(())
}

fn is_ignored(name: &str) -> bool {
    name.starts_with('.') || IGNORE.contains(&name)
}

fn load_gitignore(root: &Path) -> Vec<String> {
    let path = root.join(".gitignore");
    let Ok(raw) = fs::read_to_string(path) else {
        return Vec::new();
    };
    raw.lines()
        .map(str::trim)
        .filter(|l| !l.is_empty() && !l.starts_with('#'))
        .map(|l| l.trim_start_matches('/').to_string())
        .collect()
}

fn gitignore_match(rel: &str, pattern: &str) -> bool {
    let rel = rel.trim_start_matches('/');
    if pattern.contains('*') {
        if let Some(suffix) = pattern.strip_prefix('*') {
            return rel.ends_with(suffix);
        }
        if let Some(prefix) = pattern.strip_suffix('*') {
            return rel.starts_with(prefix);
        }
    }
    if rel == pattern || rel.ends_with(&format!("/{pattern}")) {
        return true;
    }
    rel.split('/').any(|seg| seg == pattern)
}

fn is_gitignored(rel: &str, patterns: &[String]) -> bool {
    patterns.iter().any(|p| gitignore_match(rel, p))
}

fn normalize(path: &Path) -> PathBuf {
    let mut out = PathBuf::new();
    for comp in path.components() {
        match comp {
            Component::ParentDir => {
                out.pop();
            }
            Component::CurDir => {}
            other => out.push(other.as_os_str()),
        }
    }
    out
}

pub fn resolve_safe(root: &Path, rel: &str) -> Result<PathBuf, String> {
    let cleaned = rel.replace('\\', "/");
    let cleaned = cleaned.trim_start_matches('/');
    let abs = normalize(&root.join(cleaned));
    let root_n = normalize(root);
    if abs != root_n && !abs.starts_with(&root_n) {
        return Err("Path escapes vault directory".into());
    }
    Ok(abs)
}

fn rel_posix(root: &Path, abs: &Path) -> String {
    abs.strip_prefix(root)
        .unwrap_or(abs)
        .to_string_lossy()
        .replace('\\', "/")
}

fn list_tree_at(root: &Path, dir: &Path, gitignore: &[String]) -> Result<Vec<TreeNode>, String> {
    let mut nodes = Vec::new();
    let mut entries: Vec<_> = fs::read_dir(dir)
        .map_err(|e| e.to_string())?
        .filter_map(|e| e.ok())
        .collect();
    entries.sort_by_key(|e| e.file_name());

    for entry in entries {
        let name = entry.file_name().to_string_lossy().to_string();
        if is_ignored(&name) {
            continue;
        }
        let abs = entry.path();
        let rel = rel_posix(root, &abs);
        if is_gitignored(&rel, gitignore) {
            continue;
        }
        let ft = entry.file_type().map_err(|e| e.to_string())?;
        if ft.is_dir() {
            nodes.push(TreeNode {
                name,
                path: rel,
                node_type: "folder".into(),
                children: Some(list_tree_at(root, &abs, gitignore)?),
            });
        } else {
            nodes.push(TreeNode {
                name,
                path: rel,
                node_type: "file".into(),
                children: None,
            });
        }
    }

    nodes.sort_by(|a, b| match (a.node_type.as_str(), b.node_type.as_str()) {
        ("folder", "file") => std::cmp::Ordering::Less,
        ("file", "folder") => std::cmp::Ordering::Greater,
        _ => a.name.to_lowercase().cmp(&b.name.to_lowercase()),
    });
    Ok(nodes)
}

fn list_files_at(
    root: &Path,
    dir: &Path,
    gitignore: &[String],
    out: &mut Vec<String>,
) -> Result<(), String> {
    for entry in fs::read_dir(dir).map_err(|e| e.to_string())? {
        let entry = entry.map_err(|e| e.to_string())?;
        let name = entry.file_name().to_string_lossy().to_string();
        if is_ignored(&name) {
            continue;
        }
        let abs = entry.path();
        let rel = rel_posix(root, &abs);
        if is_gitignored(&rel, gitignore) {
            continue;
        }
        let ft = entry.file_type().map_err(|e| e.to_string())?;
        if ft.is_dir() {
            list_files_at(root, &abs, gitignore, out)?;
        } else {
            out.push(rel);
        }
    }
    Ok(())
}

#[tauri::command]
pub fn get_vault_path(state: tauri::State<'_, Arc<VaultState>>) -> Result<String, String> {
    Ok(root_of(&state)?.to_string_lossy().to_string())
}

#[tauri::command]
pub fn set_vault_path(state: tauri::State<'_, Arc<VaultState>>, path: String) -> Result<String, String> {
    let requested = PathBuf::from(path.trim());
    if !requested.is_dir() {
        return Err("That folder does not exist".into());
    }

    let current = root_of(&state)?;
    if same_path(&requested, &current) {
        return Ok(current.to_string_lossy().to_string());
    }

    let canonical = fs::canonicalize(&requested).map_err(|e| e.to_string())?;
    if canonical == current {
        return Ok(current.to_string_lossy().to_string());
    }

    set_root(&state, canonical.clone())?;
    save_persisted_vault_path(&canonical)?;
    Ok(canonical.to_string_lossy().to_string())
}

#[tauri::command]
pub fn abs_path(state: tauri::State<'_, Arc<VaultState>>, path: String) -> Result<String, String> {
    let root = root_of(&state)?;
    let abs = resolve_safe(&root, &path)?;
    Ok(abs.to_string_lossy().to_string())
}

#[tauri::command]
pub async fn pick_vault_folder() -> Result<Option<String>, String> {
    Ok(rfd::AsyncFileDialog::new()
        .set_title("Choose vault folder")
        .pick_folder()
        .await
        .map(|p| p.path().to_string_lossy().to_string()))
}

#[tauri::command]
pub fn reveal_path(state: tauri::State<'_, Arc<VaultState>>, path: String) -> Result<(), String> {
    let root = root_of(&state)?;
    let abs = resolve_safe(&root, &path)?;
    if !abs.exists() {
        return Err("Path not found".into());
    }
    #[cfg(target_os = "macos")]
    {
        std::process::Command::new("open")
            .arg("-R")
            .arg(&abs)
            .spawn()
            .map_err(|e| e.to_string())?;
    }
    #[cfg(target_os = "windows")]
    {
        std::process::Command::new("explorer")
            .arg(format!("/select,{}", abs.display()))
            .spawn()
            .map_err(|e| e.to_string())?;
    }
    #[cfg(target_os = "linux")]
    {
        let parent = abs.parent().unwrap_or(&abs);
        std::process::Command::new("xdg-open")
            .arg(parent)
            .spawn()
            .map_err(|e| e.to_string())?;
    }
    Ok(())
}

#[tauri::command]
pub fn list_tree(state: tauri::State<'_, Arc<VaultState>>) -> Result<Vec<TreeNode>, String> {
    let root = root_of(&state)?;
    fs::create_dir_all(&root).map_err(|e| e.to_string())?;
    let gitignore = load_gitignore(&root);
    list_tree_at(&root, &root, &gitignore)
}

#[tauri::command]
pub fn list_files(state: tauri::State<'_, Arc<VaultState>>) -> Result<Vec<String>, String> {
    let root = root_of(&state)?;
    fs::create_dir_all(&root).map_err(|e| e.to_string())?;
    let gitignore = load_gitignore(&root);
    let mut out = Vec::new();
    list_files_at(&root, &root, &gitignore, &mut out)?;
    Ok(out)
}

#[tauri::command]
pub fn read_file(state: tauri::State<'_, Arc<VaultState>>, path: String) -> Result<String, String> {
    let root = root_of(&state)?;
    let abs = resolve_safe(&root, &path)?;
    fs::read_to_string(abs).map_err(|e| e.to_string())
}

#[tauri::command]
pub fn write_file(
    state: tauri::State<'_, Arc<VaultState>>,
    path: String,
    content: String,
) -> Result<WriteResult, String> {
    let root = root_of(&state)?;
    let abs = resolve_safe(&root, &path)?;
    let created = !abs.exists();
    if let Some(parent) = abs.parent() {
        fs::create_dir_all(parent).map_err(|e| e.to_string())?;
    }
    fs::write(&abs, content).map_err(|e| e.to_string())?;
    Ok(WriteResult { created })
}

#[tauri::command]
pub fn delete_path(state: tauri::State<'_, Arc<VaultState>>, path: String) -> Result<(), String> {
    let root = root_of(&state)?;
    let abs = resolve_safe(&root, &path)?;
    let meta = fs::metadata(&abs).map_err(|e| e.to_string())?;
    if meta.is_dir() {
        fs::remove_dir_all(abs).map_err(|e| e.to_string())
    } else {
        fs::remove_file(abs).map_err(|e| e.to_string())
    }
}

#[tauri::command]
pub fn move_path(
    state: tauri::State<'_, Arc<VaultState>>,
    from: String,
    to: String,
) -> Result<String, String> {
    let from = from.trim_start_matches('/').replace('\\', "/");
    let to = to.trim_start_matches('/').replace('\\', "/");
    if from.is_empty() || to.is_empty() {
        return Err("from and to are required".into());
    }
    if from == to {
        return Ok(to);
    }

    let root = root_of(&state)?;
    let from_abs = resolve_safe(&root, &from)?;
    let to_abs = resolve_safe(&root, &to)?;

    if !from_abs.exists() {
        return Err("Source not found".into());
    }
    if to_abs.exists() {
        return Err("A file or folder already exists at the destination".into());
    }
    if to_abs == from_abs || to_abs.starts_with(&from_abs) {
        return Err("Cannot move a folder into itself".into());
    }
    if let Some(parent) = to_abs.parent() {
        fs::create_dir_all(parent).map_err(|e| e.to_string())?;
    }
    if from_abs.is_dir() {
        move_dir(&from_abs, &to_abs)?;
    } else {
        fs::rename(&from_abs, &to_abs).map_err(|e| e.to_string())?;
    }
    Ok(to)
}

fn move_dir(from: &Path, to: &Path) -> Result<(), String> {
    if fs::rename(from, to).is_ok() {
        return Ok(());
    }
    copy_dir_all(from, to)?;
    fs::remove_dir_all(from).map_err(|e| e.to_string())
}

fn copy_dir_all(from: &Path, to: &Path) -> Result<(), String> {
    fs::create_dir_all(to).map_err(|e| e.to_string())?;
    for entry in fs::read_dir(from).map_err(|e| e.to_string())? {
        let entry = entry.map_err(|e| e.to_string())?;
        let src = entry.path();
        let dest = to.join(entry.file_name());
        if entry.file_type().map_err(|e| e.to_string())?.is_dir() {
            copy_dir_all(&src, &dest)?;
        } else {
            fs::copy(&src, &dest).map_err(|e| e.to_string())?;
        }
    }
    Ok(())
}
